espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SourceDocument.java

```diff
@@ -0,0 +1,138 @@
+    }
+
+    /**
+     * Recursively visit all of the views and map them to XML elements
+     * @param doc XML Document
+     * @param parentElement Element that this new element will be appended to
+     * @param view Android View that will map to an Element
+     */
+    private void buildXML(Document doc, Element parentElement, View view) {
+        // TODO: Investigate if this is safe and if we need to strip out characters before constructing XML document
+        Element element = doc.createElement(view.getClass().getName());
+
+        // Set attributes
+        ViewElement viewElement = new ViewElement(view);
+        element.setAttribute(ViewAttributesEnum.CONTENT_DESC.getName(), viewElement.getContentDescription().toString());
+        element.setAttribute(ViewAttributesEnum.BOUNDS.getName(), viewElement.getBounds().toShortString());
+        element.setAttribute(ViewAttributesEnum.FOCUSED.getName(), viewElement.isFocused() ? "true" : "false");
+        element.setAttribute(ViewAttributesEnum.CLICKABLE.getName(), viewElement.isClickable() ? "true" : "false");
+        element.setAttribute(ViewAttributesEnum.LONG_CLICKABLE.getName(), viewElement.isLongClickable() ? "true" : "false");
+        element.setAttribute(ViewAttributesEnum.CLASS.getName(), viewElement.getClassName());
+        element.setAttribute(ViewAttributesEnum.INDEX.getName(), Integer.toString(viewElement.getIndex()));
+
+        // If this is the rootElement, append it to the document
+        if (parentElement == null) {
+            doc.appendChild(element);
+        } else {
+            parentElement.appendChild(element);
+        }
+
+        // If cacheElementReferences == true, then cache a reference to the View
+        if (cacheElementReferences) {
+            viewMap.put(element, view);
+        }
+
+        // Visit the children and build them too
+        try {
+            for (int index = 0; index < ((ViewGroup) view).getChildCount(); ++index) {
+                View childView = ((ViewGroup) view).getChildAt(index);
+                buildXML(doc, element, childView);
+            }
+        } catch (ClassCastException e) {
+            // If it couldn't be cast to a ViewGroup, it has no children
+        }
+    }
+
+    public String toXMLString() throws TransformerException {
+        DOMSource source = new DOMSource(doc);
+        StringWriter stringWriter = new StringWriter();
+        StreamResult result = new StreamResult(stringWriter);
+        transformer.transform(source, result);
+        return stringWriter.toString();
+    }
+
+    public static View findViewByXPath(String xpathSelector) throws XPathLookupException {
+        try {
+            SourceDocument sourceDocument = new SourceDocument(true);
+            XPath xpath = XPathFactory.newInstance().newXPath();
+            Element elementNode = (Element) xpath.evaluate(xpathSelector, sourceDocument.doc, XPathConstants.NODE);
+            if (elementNode != null)
```

Comment:
it is recommended to always encode condition body into {} even if there is only open operator.